modules/utxo-lib/src/bitgo/UtxoPsbt.ts

```diff
@@ -317,3 +376,145 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
+        if (!publicKeys || publicKeys.length !== 2) {
+          throw new Error('expected 2 pubkeys');
+        }
+        if (witness.length !== 2) {
+          throw new Error(`expected exactly 2 signatures, got ${witness.length}`);
+        }
+        update.tapScriptSig = [];
+        let signature;
+        while ((signature = witness.pop()) !== undefined) {
+          if (signature.length === 0) {
+            publicKeys.shift(); // No signature for this key
+            continue;
+          }
+          const leafHash = taproot.getTapleafHash(eccLib, controlBlock, script);
+          const pubkey = publicKeys.shift();
+          if (!pubkey) throw new Error("Impossible, known 2-length things didn't match");
+          update.tapScriptSig.push({ signature, pubkey, leafHash });
+        }
+      } else if (witnessProgramCandidate.length === 34 && witnessProgramCandidate[0] === opcodes.OP_0) {
+        // P2WSH
+        const witnessScript = witness.pop();
+        if (!witnessScript) throw new Error('Invalid witness structure');
+        update.witnessScript = witnessScript;
+        const decompiledWitnessScript = bscript.decompile(witnessScript);
+        if (!decompiledWitnessScript) {
+          throw new Error('Invalid witnessScript, failed to decompile');
+        }
+        const hashFn = (hashType) => tx.hashForWitnessV0(vin, witnessScript, prevOut.value, hashType);
+        update.partialSig = matchSignatures(hashFn, witnessScript, witness);
+      } else if (witnessProgramCandidate.length === 22 && witnessProgramCandidate[0] === opcodes.OP_0) {
+        // P2WPKH
+        if (witness.length === 2) {
+          const pubkey = witness.pop();
+          const signature = witness.pop();
+          if (!Buffer.isBuffer(pubkey) || !Buffer.isBuffer(signature)) {
+            throw new Error('Invalid pubkey or signature');
+          }
+          update.partialSig = [{ pubkey, signature }];
+        }
+        return update;
+      }
+    }
+    return update;
+  });
+  tx.ins.forEach((input) => {
+    input.witness = [];
+    input.script = Buffer.alloc(0);
+  });
+  return ret;
+}
+
+function matchSignatures(hashFn: (hashType: number) => Buffer, script: Buffer, stack: Stack): PartialSig[] {
+  const partialSig: PartialSig[] = [];
+  const publicKeys = payments.p2ms({ output: script }).pubkeys;
+  if (!publicKeys || publicKeys.length !== 3) {
+    throw new Error('Invalid multisig script');
+  }
+  for (const sig of stack.slice(1)) {
+    // Ignore extra empty element re MULTISIG bug
+    if (!Buffer.isBuffer(sig) || sig.length === 0) continue;
+    const { signature, hashType } = ScriptSignature.decode(sig);
+    const hash = hashFn(hashType);
+    let signatureMatched = false;
+    for (const pubkey of publicKeys) {
+      if (!eccLib.verify(hash, pubkey, signature)) continue;
+      signatureMatched = true;
+      partialSig.push({ pubkey, signature: sig });
```

Comment:
Fwiw to me this is slightly easier to follow without the `continue` statement and the negated `if`, e.g
```suggestion
    for (const pubkey of publicKeys) {
      if (eccLib.verify(hash, pubkey, signature)) {
        signatureMatched = true;
        partialSig.push({ pubkey, signature: sig });
        break;
      }
```

Reply:
Yeah, actually just made that same change in response to your above comment :).